src/main/java/jewellery/inventory/service/SaleService.java

```diff
@@ -33,23 +36,53 @@ public SaleResponseDto createSale(SaleRequestDto saleRequestDto) {
             userService.getUser(saleRequestDto.getBuyerId()),
             getProductsFromSaleRequestDto(saleRequestDto));
 
-    throwExceptionIfProductOwnerNotSeller(sale.getProducts(), saleRequestDto.getSellerId());
+    throwExceptionIfProductIsSold(sale.getProducts());
+    throwExceptionIfUserNotSeller(sale.getProducts(), saleRequestDto.getSellerId());
+    throwExceptionIfProductIsPartOfAnotherProduct(sale.getProducts());
+
     Sale createdSale = saleRepository.save(sale);
     updateProductOwnersAndSale(sale.getProducts(), saleRequestDto.getBuyerId(), createdSale);
     return saleMapper.mapEntityToResponseDto(createdSale);
   }
 
-  private void throwExceptionIfProductOwnerNotSeller(List<Product> products, UUID sellerId) {
+  private void throwExceptionIfUserNotSeller(List<Product> products, UUID sellerId) {
     for (Product product : products) {
       if (!product.getOwner().getId().equals(sellerId)) {
         throw new UserNotOwnerException(product.getOwner().getId(), sellerId);
       }
     }
   }
 
+  private void throwExceptionIfProductIsPartOfAnotherProduct(List<Product> products) {
+    for (Product product : products) {
+      if (product.getContentOf() != null) {
+        throw new ProductIsContentException(product.getId());
+      }
+    }
+  }
+
+  private void throwExceptionIfProductIsSold(List<Product> products) {
+    for (Product product : products) {
+      if (product.getPartOfSale() != null) {
+        throw new ProductIsSoldException(product.getId());
+      }
+    }
+  }
+
   private void updateProductOwnersAndSale(List<Product> products, UUID buyerId, Sale sale) {
+    User newOwner = userService.getUser(buyerId);
     for (Product product : products) {
-      productService.updateProductOwnerAndSale(product, userService.getUser(buyerId), sale);
+      updateOwnerAndSubProducts(product, newOwner, sale);
+    }
+  }
+
+  private void updateOwnerAndSubProducts(Product product, User newOwner, Sale sale) {
+    productService.updateProductOwnerAndSale(product, newOwner, sale);
+    if (product.getProductsContent() != null) {
+      List<Product> subProducts = product.getProductsContent();
+      for (Product subProduct : subProducts) {
+        updateOwnerAndSubProducts(subProduct, newOwner, sale);
+      }
```

Comment:
We don't need this? We moved the logic in ProductService.

Reply:
Removed